tests/Authentication/Authenticators/HmacAuthenticatorTest.php

```diff
@@ -0,0 +1,263 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Authentication\Authenticators;
+
+use CodeIgniter\I18n\Time;
+use CodeIgniter\Shield\Authentication\Authentication;
+use CodeIgniter\Shield\Authentication\Authenticators\HmacSha256;
+use CodeIgniter\Shield\Config\Auth;
+use CodeIgniter\Shield\Entities\AccessToken;
+use CodeIgniter\Shield\Entities\User;
+use CodeIgniter\Shield\Models\UserIdentityModel;
+use CodeIgniter\Shield\Models\UserModel;
+use CodeIgniter\Shield\Result;
+use CodeIgniter\Test\Mock\MockEvents;
+use Config\Services;
+use Tests\Support\DatabaseTestCase;
+
+/**
+ * @internal
+ */
+final class HmacAuthenticatorTest extends DatabaseTestCase
+{
+    private HmacSha256 $auth;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $config = new Auth();
+        $auth   = new Authentication($config);
+        $auth->setProvider(model(UserModel::class));
+
+        Config('AuthToken')->recordLoginAttempt = Auth::RECORD_LOGIN_ATTEMPT_ALL;
+
+        /** @var HmacSha256 $authenticator */
+        $authenticator = $auth->factory('hmac');
+        $this->auth    = $authenticator;
+
+        Services::injectMock('events', new MockEvents());
+    }
+
+    public function testLogin(): void
+    {
+        $user = fake(UserModel::class);
+
+        $this->auth->login($user);
+
+        // Stores the user
+        $this->assertInstanceOf(User::class, $this->auth->getUser());
+        $this->assertSame($user->id, $this->auth->getUser()->id);
+    }
+
+    public function testLogout(): void
+    {
+        // this one's a little odd since it's stateless, but roll with it...
+        $user = fake(UserModel::class);
+
+        $this->auth->login($user);
+        $this->assertNotNull($this->auth->getUser());
+
+        $this->auth->logout();
+        $this->assertNull($this->auth->getUser());
+    }
+
+    public function testLoginByIdNoToken(): void
+    {
+        $user = fake(UserModel::class);
```

Comment:
```suggestion
        /** @var User $user */
        $user = fake(UserModel::class);
```

Reply:
@datamweb Would not have thought that would have made a difference.  I have learned something new.  Thank you!